### PR TITLE
[Test Harness] Add test harness library

### PR DESCRIPTION
**Pull request by alexvanyo**
Adds a new testing library, `testharness` for testing Compose layouts.

### CODE REVIEW

**Comment by yschimke** [comment]
Want to add isRound? :)

**Review by bentrengrove — CHANGES_REQUESTED** [review]
Please add an issue template to .github/ISSUE_TEMPLATE and also an entry for the docs generation into mkdocs.yml

Also, is there no sample for this project?

**Comment by alexvanyo** [comment]
> Want to add isRound? :)

For `isRound`, I'd want to do more investigation with how the APIs handle rendering outside of the round viewport on a circular watch.

For example: if you render something in the very corner of a physically round watch (such that it isn't actually visible), do the testing APIs treat it as being visible?

If they do, the `isRound` should be just a matter of flipping the `Configuration` flag, otherwise we could inset the content within the round viewport.

**Comment by alexvanyo** [comment]
> Please add an issue template to .github/ISSUE_TEMPLATE and also an entry for the docs generation into mkdocs.yml

Added, along with a first draft of documentation as well.

> Also, is there no sample for this project?

There are the tests of the `TestHarness` itself, and I converted some tests in `pager` and `adaptive` where it makes sense to use `TestHarness`, but agreed there should be more.

Do you think it makes sense to add some tests to `sample` that make use of it? I was thinking a navigation rail versus navigation bar test setup would be a good showcase of it.

**Review by bentrengrove — APPROVED** [review]
LGTM! Amazing work.

For the sample app, if you don't want to add a test in the sample app, just a note in your docs about where to see a sample in the repo would be good.

**Comment by yschimke** [comment]
> > Want to add isRound? :)
> 
> For `isRound`, I'd want to do more investigation with how the APIs handle rendering outside of the round viewport on a circular watch.
> 
> For example: if you render something in the very corner of a physically round watch (such that it isn't actually visible), do the testing APIs treat it as being visible?
> 
> If they do, the `isRound` should be just a matter of flipping the `Configuration` flag, otherwise we could inset the content within the round viewport.

Yep, I can follow up.

The answer is yes, they can draw outside the bounds and screenshot tools show content outside.  So for Paparazzi I change both the configuration as you suggest, but also apply a circular crop.

**Comment by JoseAlcerreca** [comment]
Given the sample is a catalog, we could add a testharness sample (with a big warning so that it's not used in prod)

image

Once I fix the darkMode issue...

Thoughts?

**Comment by JoseAlcerreca** [comment]
I've pushed it anyway, but happy to revert.

**Comment by yschimke** [comment]
I tried this to see if it could replace my bespoke code.  Not sure if it's AGP or paparazzi related (I doubt given it's debug code, not test). But seeing

```
java.lang.NullPointerException: Cannot invoke "android.content.Context.getResources()" because "resContext" is null   at android.view.ContextThemeWrapper.getResourcesInternal(ContextThemeWrapper.java:137)   at android.view.ContextThemeWrapper.getResources(ContextThemeWrapper.java:128)   at androidx.compose.ui.res.Resources_androidKt.resources(Resources.android.kt:33)   at androidx.compose.ui.res.StringResources_androidKt.stringResource(StringResources.android.kt:35)
```

Hopefully some typo in my copy/paste/simplification. But wanted to flag.

**Comment by JoseAlcerreca** [comment]
> Hopefully some typo in my copy/paste/simplification. But wanted to flag.

I can't think of anything that the harness might be doing that might be affecting resource loading. Can you share your code?

**Comment by yschimke** [comment]
> > Hopefully some typo in my copy/paste/simplification. But wanted to flag.
> 
> I can't think of anything that the harness might be doing that might be affecting resource loading. Can you share your code?

https://github.com/yschimke/horologist/commit/155971e206ce01afa69238ba9ced67e44397c719

Don't block on this, mainly flagging in case it's a more general problem.

**Comment by yschimke** [comment]
For completeness I would suggest one of the tests being to change the locale and call stringResource() from a composable.

**Comment by JoseAlcerreca** [comment]
Done 
image

**Comment by alexvanyo** [comment]
I was trying out `TestHarness` in a `@Preview`, and ran into a similar looking error.

I believe the issue is the same, caused by [this line](https://cs.android.com/android/platform/superproject/+/master:frameworks/layoutlib/bridge/src/com/android/layoutlib/bridge/android/BridgeContext.java;l=1380;drc=857b1f380d02a7d1dab011850257d472a27370cc) in layoutlib.

`ContextThemeWrapper` [uses](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/view/ContextThemeWrapper.java;l=136;drc=ad9254cb1d45844e0a3b7470292b77d34d31e962) `Context.createConfigurationContext` to apply the override, and it looks like this isn't supported in layoutlib directly, and instead, `null` is returned which leads to the crash.

A proper fix would likely be adding some functionality to layoutlib, to support `ContextThemeWrapper`? https://issuetracker.google.com/issues/200845657 seems similar.

**Comment by alexvanyo** [comment]
List of updates since your approval @bentrengrove :
- Added test harness sample
- Added locale overriding to sample and tests
- Documented current `@Preview` limitation

**Comment by JoseAlcerreca** [comment]
Got the Trengrove :rocket: of approval offline, merging!

File: testharness/src/main/java/com/google/accompanist/testharness/TestHarness.kt
Comment (alashow) [inline testharness/src/main/java/com/google/accompanist/testharness/TestHarness.kt:109]: @alexvanyo `java.util.Locale.setDefault(locales[0])` would be also useful here since java formatters won't know about the locale update otherwise

Comment (alexvanyo) [reply testharness/src/main/java/com/google/accompanist/testharness/TestHarness.kt:109]: I'd be interested to hear your use case here, since I'd intentionally prefer to avoid calling `java.util.Locale.setDefault(locales[0])` if possible.

`java.util.Locale.getDefault()` isn't observable by Compose, so if you're calling that from a `@Composable` function, you might not be notified if the locale changes in the general case (for instance, if opting out of activity recreation for locale changes)

If possible, can you retrieve the locale from `LocalConfiguration.current.locales` instead to create the formatter?

**Pull request merged**
